### ISSUE TITLE
Issue getting compiled Linux binary to start on Manjaro

### ISSUE THREAD

**Issue by garethppls** [OpenLoco/OpenLoco#1711]
Hi all,

I'm running Manjaro and I am trying to get the code running with the Locomotion files that I have.

I compiled the code in an Ubuntu docker container and the executable seems to be able to start.
![image](https://user-images.githubusercontent.com/95966098/200405673-3b3f4e0d-2338-4e90-b2d7-f733e78c7f16.png)

I provide the folder to the Locomotion installation directory. I installed it from the CD via Steam and Proton.
**/home/tv/.steam/steam/steamapps/compatdata/3018806553/pfx/drive_c/Program Files (x86)/Atari/Locomotion**

It then returns that it cannot find g1.dat

I thought maybe the issue was with Program Files having a space in it, so I copied the folder to my home folder.

![image](https://user-images.githubusercontent.com/95966098/200407018-ce764246-6a08-4851-a5c2-6c3e32eb20a7.png)

It complains about g1.dat again. I notice the extension originally has capital .DAT
I rename it and try again.
Same error.

Any ideas as to where to go from here?

**Comment by duncanspumpkin** [comment]
It tries to find g1.dat at `"<install folder>/Data/g1.DAT"` so double check that it is there. Although it does sound like you have everything setup correctly. Could you provide the details of your config file `openloco.yml`

**Comment by garethppls** [comment]
The file is definitely there:
![image](https://user-images.githubusercontent.com/95966098/200412331-d8de578b-78ce-4efc-97de-35404a0e90eb.png)

The config is more interesting. I didn't see one of those when I grabbed the compiled file from my container. Do I need to create a config file? Or where does it normally live on the file system?

**Comment by AaronVanGeffen** [comment]
It might just be the case sensitive file system that's an issue.

**Comment by garethppls** [comment]
OK - the plot thickens. When I explicitly add /home/tv/locomotion I get an io error.
![image](https://user-images.githubusercontent.com/95966098/200414634-36faec95-1667-43f1-b41d-af9b65d87312.png)

I'll just check if there's some kind of file permission issue after copying the files over.

**Comment by janisozaur** [comment]
`~` is not a valid path, it's a bash-ism. See https://www.gnu.org/software/bash/manual/html_node/Tilde-Expansion.html

There is no need to post screenshot of text, where text suffices. Screenshots can't have easily text extracted from.

Post logs from executing `strace ./openloco`

**Comment by AaronVanGeffen** [comment]
Hmm, it should simply create that file when missing. Perhaps you could try to just `touch game.cfg` to create a dummy file? It should simply re-create the config if it's not up to par.

**Comment by garethppls** [comment]
This causes the game to exit immediately. It now seems to struggle on "1.TMP". I wasn't prompted for the location of the game files on this run.

I'll copy the log and check back tomorrow:
[latest.log](https://github.com/OpenLoco/OpenLoco/files/9955845/latest.log)

**Comment by duncanspumpkin** [comment]
It sounds heavily like it doesn't have permission to create files. 1.TMP, game.cfg, openloco.yml, these are all files the game creates.

**Comment by janisozaur** [comment]
perhaps `/home/tv/.config/OpenLoco` doesn't exist or didn't initially exist and that caused inability to create the new file?

**Comment by garethppls** [comment]
@janisozaur - correct. It definitely didn't exist. 

@duncanspumpkin - it'd be odd if it didn't have permissions to write into the home directory when I'm running the executable as the user. 

I'll try create all of these files in place later and see where I get to. 

Are there plans to build a Linux binary along with the others at some stage? This would greatly simplify things.

**Comment by duncanspumpkin** [comment]
Yes i just merged a whole host of cmake changes. I've only tested that the linux parts of it build on the ci's so if you have anything different to the ci then it might not work.

**Comment by garethppls** [comment]
I will do this later this evening. Another thought is do I need the 32bit version of libx11?

**Comment by AaronVanGeffen** [comment]
> I had to take the debug version because of the compilation error above.

You can pass `-DSTRICT=Off` to CMake to work around this. This disables the 'warnings are errors' directive.

**Comment by garethppls** [comment]
To update. I compiled the code in an Ubuntu VM (latest version) I created on my other Linux machine running Debian Sid. I changed the release profile to have STRICT disabled.

I copied the folder with the game files into the VM and the game starts as expected.
![image](https://user-images.githubusercontent.com/95966098/201218934-3bf51d99-a68e-4816-95f7-7d75c340ab91.png)

I will take this binary and copy it to my other computer running Manjaro to see if this runs there tomorrow. There's at least a glimmer of hope at this stage.

**Comment by duncanspumpkin** [comment]
Also technically that nullptr warning is valid. Its not about the lack of nullptr check on attached its about the previous `ps` being possibly nullptr.

**Comment by garethppls** [comment]
So I copied the binaries across, but we still have the same issue. I checked if the 32 bit version of libx11 was on the machine and it is.

@janisozaur - Here's the logs you requested without the folder. I.E completely deleted from .config/OpenLoco
[without-folder.log](https://github.com/OpenLoco/OpenLoco/files/9992728/without-folder.log)

To be honest - it looks like the challenge is more trying to cater for different Linux distributions. I'll try and see if I can build a Flatpak to try and sandbox it so it can run stably across distributions. Perhaps a little build project for the next few weeks.

**Comment by duncanspumpkin** [comment]
So i think the issue is that we need to create the folder `.config/OpenLoco` as otherwise it will crash when we go to save the config file. I tested this out with WSL2

**Comment by garethppls** [comment]
That's one of the issues for sure. It isn't the only issue in respect to Manjaro. I'm hoping if I can build a flatpak with the required dependencies available to it that I might be able to create something that can run in a distro agnostic way. I'll keep you posted on how I get on.

My thinking is to create a docker container that'll run the build and package process and then publish the flatpak on a repository.

**Comment by janisozaur** [comment]
Ping @rmnvgr, who maintains flatpak for OpenRCT2.

**Comment by rmnvgr** [comment]
Hey @garethppls, I don't have a copy of Locomotion but I'd be happy to help you build a Flatpak for OpenLoco.

For building 32 bits modules you can have a look at this guide from the documentation: https://docs.flatpak.org/en/latest/multiarch.html

For publishing, you don't need to have a Docker container. If you have a working manifest, just submit it to Flathub, they have builders that will take care of building and publishing the Flatpak, see here: https://github.com/flathub/flathub/wiki/App-Submission

Additionally, you can ask for guidance on the Flatpak matrix channel: https://matrix.to/#/#flatpak:matrix.org

**Comment by garethppls** [comment]
@rmnvgr - do you have a repo with the manifest you use for OpenRCT2? That'd really help me get started.

**Comment by AaronVanGeffen** [comment]
For what it's worth, building a flatpak for personal use is fine, but publishing or indeed distributing it may be problematic at this point. At the moment, we're still in the process of reimplementing the game in C++. On Linux, we therefore statically link parts of the original binary into the OpenLoco executable for interopting purposes. Until this is no longer required, I think such binaries should not be redistributed.

**Comment by garethppls** [comment]
I'll start with it to see if I can get it working convincingly on my system first. I just think we need to do _something_ to provide a first-class Linux experience for this project. 

Having a binary that simply runs on whatever Linux machine I put it on is a key part of that. I guess I could use the Windows version via Proton, but that doesn't sit nicely with me if we can have a native Linux version working.

**Comment by duncanspumpkin** [comment]
Mac has no support at all at this time and won't get support until we fully reimplement the game.

**Comment by garethppls** [comment]
The code still does not run on Manjaro. Should this issue be closed?

**Comment by garethppls** [comment]
Another thing I am going to try to do today is to compare the dependencies in the Ubuntu VM with what I've got on Manjaro and fill in gaps if there are any and retry. 

If it works I'll note a list of packages here. If it doesn't I'll press on with the Flatpak (I think I'll be working on this irrespective).

**Comment by garethppls** [comment]
Got the game working via the flatpak build I created [https://github.com/garethppls/openloco-flatpak](here).

I had to make some modifications to the build files to get this far on a fork. I'll try and clean up the changes and propose a PR. 

It's pretty satisfying to have a Linux build that can be built in a consistent environment and run anywhere.

The obligatory screenshot!
[image](https://user-images.githubusercontent.com/95966098/202561713-4a266882-c31d-47db-ab6f-8ae9d18d8900.png)
